### PR TITLE
Fix issue #165. zrepl status: hide progress bar once all filesystems reach terminal state

### PR DESCRIPTION
**Pull request by 0x3333**
Hide the progress bar once all filesystems reach terminal state on pruner.

Before:
image

After:
image

### CODE REVIEW

**Comment by problame** [comment]
Thanks for fixing this

I don't get why the CI is failing.
1. Did it prompt you for anything specific? 
2. Please rebase & force-push to trigger it again.

**Comment by 0x3333** [comment]
> I don't get why the CI is failing.

I was not logged in CircleCI 🤦🏻‍♂️... rebased and forced-push. Let's see.

File: client/status/viewmodel/render.go [hunk 1]
```diff
@@ -493,15 +495,17 @@ func renderPrunerReport(t *stringbuilder.B, r *pruner.Report, fsfilter FilterFun
...
-	t.Write(">")
-	t.Write(stringbuilder.Times("-", 80-progress))
-	t.Write("]")
-	t.Printf(" %d/%d snapshots", completedDestroyCount, totalDestroyCount)
-	t.Newline()
+	if state != pruner.Done {
+		progress := int(math.Round(80 * float64(completedDestroyCount) / float64(totalDestroyCount)))
+		t.Write("Progress: ")
+		t.Write("[")
+		t.Write(stringbuilder.Times("=", progress))
```
Comment (InsanePrawn) [inline client/status/viewmodel/render.go:502]: @problame any reason why this doesn't use `t.DrawBar()` like the code above?

File: client/status/viewmodel/render.go [hunk 2]
```diff
@@ -374,13 +374,15 @@ func renderReplicationReport(t *stringbuilder.B, rep *report.Report, history *by
 			eta = time.Duration((float64(expected)-float64(replicated))/float64(rate)) * time.Second
 		}
 
-		t.Write("Progress: ")
-		t.DrawBar(50, replicated, expected, changeCount)
-		t.Write(fmt.Sprintf(" %s / %s @ %s/s", ByteCountBinaryUint(replicated), ByteCountBinaryUint(expected), ByteCountBinary(rate)))
-		if eta != 0 {
-			t.Write(fmt.Sprintf(" (%s remaining)", humanizeDuration(eta)))
+		if latest.State != report.AttemptDone {
```
Comment (problame) [inline client/status/viewmodel/render.go:377]: States `AttemptFanOutError` and `AttemptPlanningError` are also terminal states where displaying a progress bar doesn't make sense.

Consider adding an `IsTerminal()` method to `AttemptState`

Here's the code that sets the state:

https://github.com/zrepl/zrepl/blob/b9250a41a25796969246710e55c972f24d1284d6/replication/driver/replication_driver.go#L698-L715

File: client/status/viewmodel/render.go [hunk 3]
```diff
@@ -493,15 +495,17 @@ func renderPrunerReport(t *stringbuilder.B, r *pruner.Report, fsfilter FilterFun
...
-	progress := int(math.Round(80 * float64(completedDestroyCount) / float64(totalDestroyCount)))
-	t.Write("Progress: ")
-	t.Write("[")
-	t.Write(stringbuilder.Times("=", progress))
-	t.Write(">")
-	t.Write(stringbuilder.Times("-", 80-progress))
-	t.Write("]")
-	t.Printf(" %d/%d snapshots", completedDestroyCount, totalDestroyCount)
-	t.Newline()
+	if state != pruner.Done {
```
Comment (problame) [inline client/status/viewmodel/render.go:498]: Similarly, for pruner, terminal states are  `PlanErr`, `ExecErr` and `Done`

Learn more by checking the places where `doOneAttempt` sets the state and returns

https://github.com/zrepl/zrepl/blob/1f0f2f856966713c06b76943d9de0c560f820557/daemon/pruner/pruner.go#L352

**Review by problame — CHANGES_REQUESTED** [review]
Thanks for working on this. I think we should also hide the status bar in terminal error states, not just in terminal success states. WDYT?

File: client/status/viewmodel/render.go (hunk 1 above)
Comment (problame) [reply client/status/viewmodel/render.go:502]: Just an oversight, I think. @0x3333, feel free to fix this in this PR.

Comment (0x3333) [reply client/status/viewmodel/render.go]: I just added the `if`(😅), but I updated it to use DrawBar. I used `changeCount` equal `1` so the progress doesn't show as stalled.

File: client/status/viewmodel/render.go (hunk 2 above)
Comment (0x3333) [reply client/status/viewmodel/render.go]: I added the `IsTerminal` method. Nice catch

File: client/status/viewmodel/render.go (hunk 3 above)
Comment (0x3333) [reply client/status/viewmodel/render.go]: Same as `AttemptState`.

**Pull request merged**